internal/service/storage/storage.go

```diff
@@ -2,448 +2,577 @@ package storage
+		resp.Diagnostics.Append(diags...)
+	}
+	resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
+}
 
-		_ = d.Set("import", importBlock)
+func (r *storageResource) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+	var data storageModel
+	resp.Diagnostics.Append(req.State.Get(ctx, &data)...)
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	if data.ID.ValueString() == "" {
+		resp.State.RemoveResource(ctx)
+		return
+	}
+
+	storage, err := r.client.GetStorageDetails(ctx, &request.GetStorageDetailsRequest{
+		UUID: data.ID.ValueString(),
+	})
+	if err != nil {
+		if utils.IsNotFoundError(err) {
+			resp.State.RemoveResource(ctx)
+		} else {
+			resp.Diagnostics.AddError(
+				"Unable to read storage details",
+				utils.ErrorDiagnosticDetail(err),
+			)
+		}
+		return
 	}
 
-	return diags
+	resp.Diagnostics.Append(setValues(ctx, &data, storage)...)
+	resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
 }
 
-func resourceStorageUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
-	client := meta.(*service.Service)
-	diags := diag.Diagnostics{}
+func (r *storageResource) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
+	var data, state storageModel
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &data)...)
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
 
-	_, err := client.WaitForStorageState(ctx, &request.WaitForStorageStateRequest{
-		UUID:         d.Id(),
+	storage, err := r.client.WaitForStorageState(ctx, &request.WaitForStorageStateRequest{
+		UUID:         data.ID.ValueString(),
 		DesiredState: upcloud.StorageStateOnline,
 	})
 	if err != nil {
-		return diag.FromErr(err)
+		resp.Diagnostics.AddError(
+			"Storage did not reach online state",
+			utils.ErrorDiagnosticDetail(err),
+		)
+		return
 	}
 
-	req := request.ModifyStorageRequest{
-		UUID:  d.Id(),
-		Size:  d.Get("size").(int),
-		Title: d.Get("title").(string),
+	var labels map[string]string
+	if !data.Labels.IsNull() && !data.Labels.IsUnknown() {
+		resp.Diagnostics.Append(data.Labels.ElementsAs(ctx, &labels, false)...)
+	}
+	labelsSlice := utils.NilAsEmptyList(utils.LabelsMapToSlice(labels))
+
+	backupRule, diags := buildBackupRule(ctx, data.BackupRule)
+	resp.Diagnostics.Append(diags...)
+	apiReq := request.ModifyStorageRequest{
+		BackupRule: backupRule,
+		Labels:     &labelsSlice,
+		Size:       int(data.Size.ValueInt64()),
+		Title:      data.Title.ValueString(),
+		UUID:       data.ID.ValueString(),
 	}
 
-	if d.HasChange("backup_rule") {
-		if br, ok := d.GetOk("backup_rule.0"); ok {
-			backupRule := BackupRule(br.(map[string]interface{}))
-			if backupRule.Interval == "" {
-				req.BackupRule = &upcloud.BackupRule{}
-			} else {
-				req.BackupRule = backupRule
-			}
+	// Attached server must be stopped before resizing the storage
+	sizeChanged := data.Size.ValueInt64() != state.Size.ValueInt64()
```

Comment:
There seems to be a helper method for this as well (not a blocker).

```suggestion
	sizeChanged := data.Size.Equal(state.Size)
``